Template10 (Library)/Services/NavigationService/NavigationService.cs

```diff
@@ -290,8 +299,16 @@ public void Navigate<T>(T key, object parameter = null, NavigationTransitionInfo
             state.Write<string>("NavigateState", FrameFacadeInternal?.NavigationService.NavigationState);
         }
 
-        public event TypedEventHandler<Type> AfterRestoreSavedNavigation;
-        public async Task<bool> RestoreSavedNavigationAsync()
+        public event TypedEventHandler<Type> OnNavigationStateLoaded;
```

Comment:
I have created for you a problem, @callummoffat. You have built it exactly the way I asked you to do it. But 1. if the developer overrides the method with their own func<> then they will NOT be able to call the event because outside code cannot raise an inside event. There are two possible solutions. The first is to expose a Raise() method as public that the func<> could call. I don't like this because it allows for the event to be raised when it is inappropriate. Not likely, I know, but needlessly possible. The second fix is to wrap the call with an internal method like this:

````csharp
public event EventHandler BeforeSave;
void RaiseBeforeSave() => BeforeSave?.Invoke(this, new EventArgs { });

public event EventHandler AfterSave;
void RaiseAfterSave() => AfterSave?.Invoke(this, new EventArgs { });

async void CallSaveAsync()
{
    var args = new System.ComponentModel.CancelEventArgs();
    BeforeSave?.Invoke(this, args);
    if (!args.Cancel)
    {
        await SaveAsync();
        AfterSave?.Invoke(this, EventArgs.Empty);
    }
}

private Func<Task> _saveAsync;
public Func<Task> SaveAsync
{
    get { return _saveAsync ?? DefaultSaveImplementationAsync; }
    set { _saveAsync = value; }
}
private async Task DefaultSaveImplementationAsync()
{
    await Task.CompletedTask;
}
````

This is a lot for a little, huh? I think it's marvelously flexible, but I am not sure it is necessary for 1.1.12. It makes most sense for 2.0x. As a result, @callummoffat, and I am sorry for asking you to do this, but let's revert to simple Save() and Load() methods. There's no good reason to introduce this pattern when no other method is like it. Make sense? I really led you down the wrong road with this and I apologize.